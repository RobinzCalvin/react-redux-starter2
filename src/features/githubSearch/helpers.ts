export function injectSizeToAvatarURL(URL: string, sizeInPixels: number) {
  return `${URL}&s=${sizeInPixels}`;
}
