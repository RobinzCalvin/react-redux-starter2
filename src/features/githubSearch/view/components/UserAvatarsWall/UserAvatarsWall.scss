.users-avatars-wall {
  &__avatars {
    position: relative;
    display: flex;
    flex-wrap: wrap;
    align-content: flex-start;
    padding: 0;
    list-style: none;

    &_size {
      &_big {
        min-height: 24rem;
        max-width: 90rem;
      }

      &_medium {
        min-height: 28rem;
        max-width: 45rem;
      }

      &_small {
        min-height: 14rem;
        max-width: 45rem;
      }
    }
  }

  &__avatar {
    margin: 0.25rem;
    width: 4rem;
    height: 4rem;
    cursor: pointer;
    transition: transform 0.15s;

    &:hover {
      transform: scale(1.15);
    }
  }

  &__image {
    width: 100%;
    height: 100%;
  }
}
