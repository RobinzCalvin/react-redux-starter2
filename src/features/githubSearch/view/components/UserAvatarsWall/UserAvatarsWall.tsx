import * as React from 'react';
import { bind } from 'decko';

import { IUser } from 'shared/types/models';

interface IProps {
  users: IUser[];
  onAvatarClick(user: IUser): void;
}

class UserAvatarsWall extends React.PureComponent<IProps> {
  public render() {
    const { users } = this.props;
    return (
      <div className="user-avatars-wall" style={{ cursor: 'pointer' }}>
        <ul
          className="avatars"
          style={{
            listStyle: 'none',
            padding: 0,
            display: 'flex',
            width: '90rem',
            flexWrap: 'wrap',
          }}
        >
          {users.map(this.renderUserAvatar)}
        </ul>
      </div>
    );
  }

  @bind
  private renderUserAvatar(user: IUser, i: number) {
    const { avatarURL } = user;
    return (
      <li
        key={i}
        className="avatar"
        onClick={this.makeAvatarClickHandler(user)}
        style={{ textAlign: 'center', fontSize: '0.7rem', margin: '0.25rem', width: '4rem' }}
      >
        <img src={avatarURL} style={{ width: '100%' }}/>
      </li>
    );
  }

  @bind
  private makeAvatarClickHandler(user: IUser) {
    return () => this.props.onAvatarClick(user);
  }
}

export default UserAvatarsWall;
