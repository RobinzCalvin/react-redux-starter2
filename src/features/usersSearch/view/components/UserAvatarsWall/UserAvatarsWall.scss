.users-avatars-wall {
  display: flex;
  justify-content: center;

  &__avatars {
    position: relative;
    display: flex;
    flex-wrap: wrap;
    align-content: flex-start;
    justify-content: center;
    margin: 0;
    padding: 0;
    list-style: none;

    &_size {
      &_big {
        max-width: 90rem;
      }

      &_small {
        max-width: 45rem;
      }
    }
  }

  &__avatar {
    margin: 0.25rem;
    width: 4rem;
    height: 4rem;
    cursor: pointer;
    transition: transform 0.15s;

    &:hover {
      transform: scale(1.15);
    }
  }

  &__image {
    width: 100%;
    height: 100%;
  }
}
