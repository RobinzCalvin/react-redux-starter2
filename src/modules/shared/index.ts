export { default as Layout } from './Layout/Layout';
export { default as withLayout } from './withLayout';
