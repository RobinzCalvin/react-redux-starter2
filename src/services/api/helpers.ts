import { IUserSearchOptions } from 'shared/types/github';

function constructReposNumberQuery(minRepos?: number, maxRepos?: number) {
  if (maxRepos === void 0 && (minRepos && minRepos > 0)) {
    return `+repos:>${minRepos}`;
  } else if (minRepos === void 0 && (maxRepos && maxRepos > 0)) {
    return `+repos:<${maxRepos}`;
  } else if (minRepos && maxRepos) {
    return `+repos:${minRepos}..${maxRepos}`;
  }
  return '';
}

export function constructUserSearchQuery(queryString: string, options: IUserSearchOptions, page: number = 1) {
  const { searchBy, searchType, minRepos, maxRepos, reposLanguage, perPage } = options;
  return queryString.concat(
    searchBy !== 'username-email' ? `+in:${searchBy}` : '',
    searchType !== 'both' ? `+type:${searchType}` : '',
    reposLanguage ? `+language:${reposLanguage}` : '',
    constructReposNumberQuery(minRepos, maxRepos),
    `&per_page=${perPage}`,
    `&page=${page}`,
  );
}

/*
  header format (requesting page=7):
  <https://api.github.com/search/users?q=asd&per_page=30&page=6>; rel="prev",
  <https://api.github.com/search/users?q=asd&per_page=30&page=8>; rel="next",
  <https://api.github.com/search/users?q=asd&per_page=30&page=34>; rel="last",
  <https://api.github.com/search/users?q=asd&per_page=30&page=1>; rel="first"
*/
export function getTotalPagesFromLinkHeader(link?: string): number {
  if (link !== void 0) {
    const lastPageMatch = link.match(/&page=(\d+)>; rel="last"/);
    if (lastPageMatch) {
      return Number(lastPageMatch[1]);
    } else {
      const prevPageMatch = link.match(/&page=(\d+)>; rel="prev"/);
      if (prevPageMatch) {
        return Number(prevPageMatch[1]) + 1;
      } else {
        console.error(`Error while trying to get total pages from ${link}`);
      }
    }
  }
  return 1;
}
