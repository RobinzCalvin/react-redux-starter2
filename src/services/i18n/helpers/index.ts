export * from './guards';
export * from './buildTranslationKeys';
