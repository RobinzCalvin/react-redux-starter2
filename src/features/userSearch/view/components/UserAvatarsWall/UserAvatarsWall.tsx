import * as React from 'react';
import { bind } from 'decko';
import block from 'bem-cn';
import * as R from 'ramda';

import { IUser } from 'shared/types/models';
import { Preloader } from 'shared/view/elements';

import { injectSizeToAvatarURL } from '../../../helpers';
import './UserAvatarsWall.scss';

interface IState {
  areAllAvatarsLoaded: boolean;
}

interface IProps {
  users: IUser[];
  onAvatarClick(user: IUser): void;
}

const b = block('users-avatars-wall');

class UserAvatarsWall extends React.PureComponent<IProps> {
  public state: IState = {
    areAllAvatarsLoaded: false,
  };

  private avatarSize = 70;

  private get wallSize() {
    const usersNumber = this.props.users.length;
    if (usersNumber <= 30) {
      return 'small';
    }
    if (usersNumber <= 50) {
      return 'medium';
    }
    return 'big';
  }

  private get areAllAvatarsLoaded() {
    return Object.values(this.avatarsLoadingStatus).every(x => x === true);
  }

  // { avatarURL: true/false };
  private avatarsLoadingStatus: Record<string, boolean> = (() => {
    const { users } = this.props;
    const avatarsURLs = users.map(x => x.avatarURL);
    return R.zipObj(avatarsURLs, Array(users.length).fill(false));
  })();

  public componentDidUpdate({ users: prevUsers }: IProps) {
    const { users } = this.props;
    if (users !== prevUsers) {
      const currentAvatarsURLs = users.map(x => x.avatarURL);
      const prevAvatarsURLs = prevUsers.map(x => x.avatarURL);

      const addedAvatarsURLs = R.difference(currentAvatarsURLs, prevAvatarsURLs);
      const existingAvatarsURLs = R.without(addedAvatarsURLs, currentAvatarsURLs);

      this.avatarsLoadingStatus = {
        ...R.zipObj(existingAvatarsURLs, existingAvatarsURLs.map(x => this.avatarsLoadingStatus[x])),
        ...R.zipObj(addedAvatarsURLs, Array(addedAvatarsURLs.length).fill(false)),
      };

      if (!this.areAllAvatarsLoaded) {
        this.setState({ areAllAvatarsLoaded: false });
      }
    }
  }

  public render() {
    const { users } = this.props;
    const { areAllAvatarsLoaded } = this.state;
    return users.length > 0 && (
      <div className={b()}>
        <ul className={b('avatars', { size: this.wallSize })}>
          <Preloader isShown={!areAllAvatarsLoaded} size={100} backgroundColor="#fff"/>
          {users.map(this.renderUserAvatar)}
        </ul>
      </div>
    );
  }

  @bind
  private renderUserAvatar(user: IUser, i: number) {
    const { avatarURL } = user;
    return (
      <li
        key={i}
        className={b('avatar')}
        onClick={this.makeAvatarClickHandler(user)}
      >
        <img
          className={b('image')}
          src={injectSizeToAvatarURL(avatarURL, this.avatarSize)}
          onLoad={this.makeImageOnLoadHandler(user.avatarURL)}
        />
      </li>
    );
  }

  @bind
  private makeAvatarClickHandler(user: IUser) {
    return () => this.props.onAvatarClick(user);
  }

  @bind
  private makeImageOnLoadHandler(avatarURL: string) {
    return () => {
      this.avatarsLoadingStatus = { ...this.avatarsLoadingStatus, [avatarURL]: true };

      if (this.areAllAvatarsLoaded) {
        this.setState({ areAllAvatarsLoaded: true });
      }
    };
  }
}

export default UserAvatarsWall;
