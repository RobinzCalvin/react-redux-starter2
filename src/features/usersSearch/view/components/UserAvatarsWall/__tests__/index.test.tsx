import React from 'react';
import { render } from 'enzyme';

import { getGithubUserMock } from 'shared/helpers';

import UserAvatarsWall, { IUserAvatarsWallProps } from '../UserAvatarsWall';

const props: IUserAvatarsWallProps = {
  users: Array(10).fill(getGithubUserMock()),
  onAvatarClick: jest.fn(),
};

describe('UserAvatarsWall component', () => {
  const component = render(<UserAvatarsWall {...props} />);
  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
