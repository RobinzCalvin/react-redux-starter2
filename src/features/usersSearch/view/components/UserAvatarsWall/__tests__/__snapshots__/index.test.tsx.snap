// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAvatarsWall component should match snapshot 1`] = `
<div
  class="users-avatars-wall"
>
  <ul
    class="users-avatars-wall__avatars users-avatars-wall__avatars_size_small"
  >
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      class="users-avatars-wall__avatar"
    >
      <img
        class="users-avatars-wall__image"
        src="http://the-url.com&s=70"
      />
    </li>
  </ul>
</div>
`;
